README.md

```diff
@@ -1,116 +1,268 @@
+```nvidia-smi```.
+If you don't have an Nvidia GPU, you can still run the project on a CPU, but some functionalities might be limited.
+
+## Features
+
+### Grid Search
+
+The Grid Search feature is designed to systematically explore a range of hyperparameters to find the optimal settings for our models. It evaluates the performance of the RetNet and Transformer models under various combinations of learning rates, embedding dimensions, feed-forward dimensions, sequence length, and number of heads. The goal is to identify the configuration that results in the best model performance, measured in terms of loss and training efficiency for a given architecture.
 
-For complex models, managing hyperparameters via a config file can be more convenient.
+**Code Overview:**
+
+We implement the grid search process as follows:
+
+- **Hyperparameters Tested:** Learning rates (`0.001`, `0.0005`, `0.0001`), embedding dimensions (`768`, `1024`, `1280`), feed-forward dimension (`1024`, `2048`), heads, (`4`, `8`), and sequence length (`256`, `512`) for a total of 72 unique combinations.
+- **Evaluation Metric:** The models are compared based on their test loss, with a custom function `evaluate_models` indicating which model performed better.
+- **Output:** Results are recorded in a CSV file, including each combination's average loss for both models, similarity scores, and training times.
+
+**Key Functions:**
+
+- `evaluate_models(model1, model2, model1_loss, model2_loss)`: Compares two models based on their test losses, returning a score that reflects their performance relative to each other.
+- `grid_search(config)`: Conducts the grid search over the predefined ranges of hyperparameters, training both RetNet and Transformer models with each combination and logging the results.
+
+**Usage:**
+
+To run the grid search, ensure your configuration file is correctly set up, then execute the script with the path to your config file as an argument:
 
 ```bash
-python train.py --config config.yaml
+python your_script_name.py path/to/your/config_file.yaml
+```
+
+### Hugging Face Integration
+
+This feature introduces custom models built upon the Hugging Face Transformers framework, enabling the incorporation of RetNet and Transformer architectures into a wide range of NLP tasks. Leveraging Hugging Face's `PreTrainedModel` class, we've developed `RetNetModelHF` and `TransformerModelHF` classes to seamlessly integrate with Hugging Face's ecosystem, facilitating easy model training, evaluation, and deployment.
+
+**Code Overview:**
+
+- **`RetNetModelHF`**: Implements the RetNet architecture as a subclass of `PreTrainedModel`, utilizing Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.
+- **`TransformerModelHF`**: Adapts the traditional Transformer architecture within the Hugging Face framework, following similar principles as the RetNet model for easy integration and use.
```

Comment:
Why is the wording for these two classes different? Don't they implement the TorchScale code in the same way? @DrewGalbraith

Reply:
I'm not sure and yes, they do.

Reply:
What is your suggestion on fixing this @nprisbrey?

Reply:
I would think that since this code is almost exactly the same, a different description is misleading because it implies that the code is different. I would think something like the following would be better, though once the descriptions are the same, it aids us in arriving to the questions once again of "Is this even needed in the README? Why isn't this just a doc string instead?"

**Code Overview:**

- **`RetNetModelHF`**: Implements the RetNet architecture as a subclass of `PreTrainedModel`, using Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.
- **`TransformerModelHF`**: Implements the Transformer architecture as a subclass of `PreTrainedModel`, using Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.

Reply:
I changed it to your description @nprisbrey.